src/targets/gpu/kernels/include/migraphx/kernels/scatter_reduction_modes.hpp

```diff
@@ -42,7 +42,7 @@ struct assign_add
     template <class T, class U>
     MIGRAPHX_DEVICE_CONSTEXPR void operator()(T& x, U y) const
     {
-        atomicAdd(&x, y);
+        unsafeAtomicAdd(&x, T(y));
```

Comment:
Why make this now non atomic? Is this due to the async fill added for hip calls?

Reply:
Its still atomic. Its just using "unsafe" atomics so it will use the atomic add in the hardware(otherwise this will fallback to compare-and-swap which is really slow).

Reply:
makes sense

Reply:
https://rocm.docs.amd.com/projects/HIP/en/latest/reference/kernel_language.html#unsafe-floating-point-atomic-rmw-operations

From the description it seems it can produce incorrect results in certain cases. How do we gurantee correct results ?

Reply:
@pfultz2  `unsafeAtomicAdd` HIP function it seems is only available for float and double.  Does this function require a `std::enable_if` to use different overload for integers and other dtypes ?

Reply:
> From the description it seems it can produce incorrect results in certain cases. How do we guarantee correct results when using "unsafe"atomic add here in MIGraphX ?

This is only a problem when doing atomics across devices, which we dont do. 

> unsafeAtomicAdd HIP function it seems is only available for float and double. Does this function require a std::enable_if to use different overload for integers and other dtypes ? Or perhaps do an assert.

I will probably need to add an `if constexpr` to check for that and then call the safe version. However, it would still be good to add an assert for debug runs as using the CAS is extremely slow.

Reply:
> I will probably need to add an if constexpr to check for that and then call the safe version. However, it would still be good to add an assert for debug runs as using the CAS is extremely slow. 

Can you add that ?